### ISSUE TITLE
NativeCallback fastcall and thiscall broken on x86/Windows

### ISSUE THREAD

**Issue by mrgreywater** [frida/frida#229]
Interceptor.replace in conjunction with NativeCallback frees too much stackspace after executing with the ABI 'fastcall' and 'thiscall'

Example msvc++ source to reproduce (Hit any key to execute all functions once): [FridaAbiTest.cpp](https://gist.github.com/mrgreywater/ad1d392dbd9b51a028e234f4a820e5fd)
Or as compiled binary (debug build -O0, no ASLR, with stack-checks): [download](http://www64.zippyshare.com/v/V32qd7E3/file.html)
Example Script for Frida: [FridaAbiTest.js](https://gist.github.com/mrgreywater/2a11eb29883785f1b5b1b7e3ae5f025f)

The example executable will run fine as long as Frida doesn't replace either the fastcall or thiscall function. As soon as that happens, the application crashes with a stack assertion (if those are activated in the compiler). If stack assertions are disabled, the application will usually just crash.

**Comment by oleavr** [comment]
Thanks! Wonder if `thiscall` implies that the first argument is a given, so `nargs` here is off by one:
https://github.com/frida/frida-gum/blob/c71ebd807f48433de0ae312fe93f83ba8ad3d4ad/bindings/gumjs/gumdukcore.c#L2882

I won't have bandwidth to look at this anytime soon, but I'm happy to assist in case you are willing to take a stab at it. Only build dependencies are MSVS2015, latest Node.js on your PATH, and Python.

**Comment by mrgreywater** [comment]
No, the first argument is not given, it's the instance 'this' pointer passed in ecx. If you remove one parameter, in the NativeCallback thiscall, it won't crash, but you don't have access to the third argument anymore (which is '2') in this example.
Note fastcall, which is a pretty similar calling convention has the same issue.
I'm kinda busy right now, so I doubt I'll find the time to look at a fix anytime soon, sorry.

**Comment by oleavr** [comment]
No worries. Let's keep tracking this at frida/frida-gum#191.

**Comment by 4got2think** [comment]
adding

```
#ifdef G_OS_WIN32
  switch (abi) {
	case FFI_THISCALL:
		  for (size_t argidx = 0; argidx < callback->cif.nargs; ++argidx)
		  {
			  if (argidx == 1)
				  break;
			  callback->cif.bytes -= callback->atypes[argidx]->size;
		  }
		  break;

	  case FFI_FASTCALL:
		  for (size_t argidx = 0; argidx < callback->cif.nargs; ++argidx)
		  {
			  if (argidx == 2)
				  break;
			  callback->cif.bytes -= callback->atypes[argidx]->size;
		  }
		  break;

	  default:
		  break;
  }
#endif
```


at https://github.com/frida/frida-gum/blob/master/bindings/gumjs/gumv8core.cpp#L2672 worked for me

hacky solution, you can think of something better for sure
it used to clean up the register arguments size too
don't know if it breaks any other code, it doesn't seem to

**Comment by oleavr** [comment]
This is now finally fixed in Frida 14.2.3, and libffi patches have been submitted upstream 🎉 Thanks for reporting and helping narrow this one down!
